gateway/src/services/base.ts

```diff
@@ -29,6 +29,40 @@ export const bigNumberWithDecimalToStr = (n: BigNumber, d: number): string => {
     .join('');
 };
 
+export const stringWithDecimalToBigNumber = (
+  numberStr: string,
+  d: number
+): BigNumber => {
+  const leftAndRight = numberStr.split('.');
+
+  if (leftAndRight.length === 2) {
+    const existingDecimals = leftAndRight[1];
+    if (existingDecimals.length > d) {
+      const right = leftAndRight[1].substr(0, d);
+      return BigNumber.from(leftAndRight[0] + right);
+    } else {
+      const neededZeros = d - existingDecimals.length;
+      const zeros = '0'.repeat(neededZeros);
+      return BigNumber.from(leftAndRight[0] + leftAndRight[1] + zeros);
+    }
+  }
+
+  const zeros = '0'.repeat(d);
+  return BigNumber.from(numberStr + zeros);
+};
+
+export const gasCostInEthString = (
+  gasPrice: number,
+  gasLimit: number
+): string => {
+  return bigNumberWithDecimalToStr(
+    BigNumber.from(gasPrice)
```

Comment:
This is unsafe because floating point numbers (e.g. 2.1) are actually valid gas prices. `BigNumber` on the other hand, only deals with integers.

I'd recommend doing the `gasPrice` * `gasLimit` calculation in floating point and then convert to integer first - since that part usually doesn't overflow the precision of floating point numbers. After that, convert to `BigNumber`. e.g.

`BigNumber.from(Math.ceil(gasPrice * gasLimit)).mul(BigNumber.from(1e9))`